### PR TITLE
Refactor: `Featured product` and `Featured category` classes

### PR DESCRIPTION
**Pull request by albarin**
Following the same idea as in the #6406 refactor, this PR aims to reduce code duplication in the `Featured Product` and `Featured Category` blocks.
- A new abstract class `FeaturedItem.php` with the common code is introduced.
- Both `FeaturedProduct` and `FeaturedCategory` classes extend the new `FeaturedItem` class and implement some functions with the specific code needed for each.

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing
1. Create a page and add a `Featured Category` block.
2. Save the page and check it renders correctly the default config in the frontend.
3. Edit the page and make some changes to the block (duotone, padding, overlay, focal point, fixed/repeated image etc.).
4. Save the page and check it renders correctly with the new settings.
5. Repeat steps 1-4 with the `Featured Product` block.

### CODE REVIEW

File: src/BlockTypes/FeaturedItem.php [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+	 * @var array
+	 */
+	protected $global_style_wrapper = array(
+		'text_color',
+		'font_size',
+		'border_color',
+		'border_radius',
+		'border_width',
+		'background_color',
+		'text_color',
```
Comment (sunyatasattva) [inline src/BlockTypes/FeaturedItem.php]: It seems that `text_color` is present twice.

Super extreme nitpick of mine: when order is not important, what do you think of ordering stuff in alphabetical order? It makes things easier to find and also to spot duplicates like this.

File: src/BlockTypes/FeaturedItem.php [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+	 */
+	abstract protected function get_item( $attributes );
+
+	/**
+	 * Returns the name of the featured item.
+	 *
+	 * @param \WP_Term|\WC_Product $item Item object.
+	 * @return string
+	 */
+	abstract protected function get_item_title( $item);
```
Comment (sunyatasattva) [inline src/BlockTypes/FeaturedItem.php]: Strange that PHPCS didn't complain, but there is a missing space.

File: src/BlockTypes/FeaturedItem.php [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+
+	/**
+	 * Returns the featured item image URL.
+	 *
+	 * @param \WP_Term|\WC_Product $item Item object.
+	 * @param string               $size    Image size, defaults to 'full'.
```
Comment (sunyatasattva) [inline src/BlockTypes/FeaturedItem.php]: I'm not sure about this, but are PHPDocs supposed to be aligned, and how? The description of `size` is off compared to the one above.

File: src/BlockTypes/FeaturedItem.php [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+	/**
+	 * Get the supports array for this block type.
+	 *
+	 * @see $this->register_block_type()
+	 * @return string[][];
+	 */
+	protected function get_block_type_supports() {
+		return array(
+			'color' => array(
+				'__experimentalDuotone' => ".wc-block-{$this->block_name}__background-image",
```
Comment (sunyatasattva) [inline src/BlockTypes/FeaturedItem.php]: Ok so… I have no idea why, but as far as I have seen in this repo, double quotes  are not used in favor of `.` concatenation. Personally, I really dislike that and I'd much rather have double quotes. But perhaps do you mind asking in the repo channel whether this is intentional and there is some reason behind it?

I recall asking when I first joined, here is the [very short thread](p1640364605498300-slack-C8X6Q7XQU). I think it'd be interesting to know more.

File: src/BlockTypes/FeaturedItem.php [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+	public function get_bg_styles( $attributes, $image_url ) {
+		$style = '';
+
+		if ( $attributes['isRepeated'] || $attributes['hasParallax'] ) {
+			$style .= "background-image: url($image_url);";
+		}
+
+		if ( ! $attributes['isRepeated'] ) {
+			$style .= 'background-repeat: no-repeat;';
+			$style .= 'background-size: ' . ( 'cover' === $attributes['imageFit'] ? $attributes['imageFit'] : 'auto' ) . ';';
```
Comment (sunyatasattva) [inline src/BlockTypes/FeaturedItem.php]: What do you think of extracting this ternary operator outside of the string concatenation for a bit more readability? It's not awfully unreadable, so I'd leave it to your judgement if you find it better like this.

**Review by sunyatasattva — CHANGES_REQUESTED** [review]
I'm super happy with this. I think we have managed to improve these blocks a lot with this refactor! My only comments are small nitpicks.

File: src/BlockTypes/FeaturedItem.php (hunk 1 above)
Comment (albarin) [reply src/BlockTypes/FeaturedItem.php]: Makes total sense 👌

File: src/BlockTypes/FeaturedItem.php (hunk 3 above)
Comment (albarin) [reply src/BlockTypes/FeaturedItem.php]: Yeah, I'm having some issue with autoformatting, I need to look into it 🤔

**Review by sunyatasattva — APPROVED** [review]
❤️  Nice job! Thank you for addressing my nitpicks. In the end you didn't decide to bring up the string concatenation thing, I see :)

**Pull request merged**